### PR TITLE
constant_exprt constructor

### PR DESCRIPTION
**Pull request by kroening**
- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [X] My contribution is formatted in line with CODING_STANDARD.md.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [X] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- [X] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by tautschnig — APPROVED** [review]
I would still appreciate if box about commit messages could be ticked...

File: src/util/std_expr.h [hunk 1]
```diff
@@ -4092,10 +4092,12 @@ class type_exprt:public exprt
 class constant_exprt:public exprt
 {
 public:
+  DEPRECATED("use constant_exprt(value, type) instead")
   constant_exprt():exprt(ID_constant)
   {
   }
 
+  DEPRECATED("use constant_exprt(value, type) instead")
```
Comment (smowton) [inline src/util/std_expr.h:4100]: Probably shouldn't deprecate this, a case like

```
constant_exprt e(ty);
if(x)
  e.set_value("x");
else
  e.set_value("not x");
```
seems sensible?

**Review by smowton — APPROVED** [review]
One minor quibble

File: src/util/std_expr.h (hunk 1 above)
Comment (tautschnig) [reply src/util/std_expr.h:4100]: I'm a strong proponent of RAII and would thus write this as `constant_exprt e(x ? "x" : "not x", ty);`

Comment (kroening) [reply src/util/std_expr.h:4100]: Could also be written as
```
irep_idt value;

if(x)
  value="x";
else
  value="not x";

constant_exprt(value, ty);
```

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 7c9f8bde021b8f140b56538a0a1ff7830a2c6a05).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85574657

**Pull request merged**